From 97586ea8d4e5239ede40a6c6e5030fe99fa6fd60 Mon Sep 17 00:00:00 2001
From: Steinar H. Gunderson <sesse@chromium.org>
Date: Wed, 11 Jun 2025 01:54:12 -0700
Subject: [PATCH] Add a fast-reject filter for the subject of selectors.

In a complex selector like .a .b .c .d#e, we would have bucketing
to quickly reject (not even look at) every element that didn't
match #e, and a Bloom filter to reject elements that don't have
.a, .b and .c as ancestors, but .d would need to go through the
regular selector matching (possibly easy selector matching).

However, we now have a Bloom filter on every element, containing
bits for the hashes for all classes and attributes set on that
element. We can use that for a fast-reject filter similar to the
ancestor Bloom filter; we store a bitmask on RuleData. For sites
that use multiple such selectors in the subject, this is a nice
win, and on others, it doesn't cost much, since the test is
extremely cheap (e.g., on x86, it reduces to two instructions:
TEST [mem],reg + JZ target). The downside, of course, is that we
grow RuleData by 32 bits for the filter.

Style perftest (Zen 3, LTO but no PGO):

  Initial style (µs)     Before     After    Perf      95% CI (BCa)
  =================== ========= ========= ======= =================
  ECommerce                5455      5389   +1.2%  [ +0.9%,  +1.5%]
  Encyclopedia            42259     41730   +1.3%  [ +0.3%,  +2.2%]
  Extension               55573     54875   +1.3%  [ +1.0%,  +1.6%]
  News                    21949     21808   +0.6%  [ +0.3%,  +0.9%]
  Search                   7690      7624   +0.9%  [ +0.5%,  +1.2%]
  Social1                 13661     13647   +0.1%  [ -0.2%,  +0.4%]
  Social2                  8652      8562   +1.1%  [ +0.7%,  +1.7%]
  Sports                  26752     26461   +1.1%  [ +0.2%,  +2.0%]
  Video                   17724     17532   +1.1%  [ +0.7%,  +1.5%]
  Geometric mean                            +1.0%  [ +0.7%,  +1.2%]

  Recalc style (µs)      Before     After    Perf      95% CI (BCa)
  =================== ========= ========= ======= =================
  ECommerce                2923      2925   -0.1%  [ -0.4%,  +0.3%]
  Encyclopedia            33593     33510   +0.2%  [ -0.1%,  +0.6%]
  Extension               43911     43604   +0.7%  [ +0.3%,  +1.1%]
  News                    10965     10858   +1.0%  [ +0.6%,  +1.3%]
  Search                   2822      2792   +1.1%  [ +0.6%,  +1.5%]
  Social1                  6795      6753   +0.6%  [ +0.3%,  +1.0%]
  Social2                  4182      4164   +0.4%  [ +0.0%,  +0.8%]
  Sports                  12639     12572   +0.5%  [ +0.2%,  +0.9%]
  Video                    7386      7054   +4.7%  [ +4.3%,  +5.1%]
  Geometric mean                            +1.0%  [ +0.8%,  +1.3%]

The memory profile is a bit confusing. Sports and Video are listed
as using significantly more GCAlloc, but even Sports only has ~25k
RuleData entries in all, so the growth should be ~100 kB, not 431 kB.
We will need to keep a watch on this.

  ECommerce       2402 →  2401 kB  (  -1 kB)
  Encyclopedia:  10354 → 10376 kB  ( +22 kB)
  Extension      11380 → 11383 kB  (  +3 kB)
  News:           6642 →  6649 kB  (  +7 kB)
  Search:         3686 →  3689 kB  (  +3 kB)
  Social1:        6067 →  6099 kB  (  +2 kB)
  Social2:        3282 →  3271 kB  ( -11 kB)
  Sports:        13130 → 13561 kB  (+431 kB)
  Video:         12084 → 12293 kB  (+209 kB)

Speedometer3 is neutral, presumably since it doesn't use too many
such selectors.

Change-Id: Icc95ebf16a082f857c2d6494db4b909665c5fded
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6621029
Reviewed-by: Anders Hartvoll Ruud <andruud@chromium.org>
Commit-Queue: Steinar H Gunderson <sesse@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1472268}
---

--- a/third_party/blink/renderer/core/css/element_rule_collector.cc
+++ b/third_party/blink/renderer/core/css/element_rule_collector.cc
@@ -551,15 +551,19 @@
         static_cast<wtf_size_t>(rules.size()));
   }
 
+  const Element::TinyBloomFilter element_filter =
+      context.element->AttributeOrClassBloomFilter();
+
   for (const RuleData& rule_data : rules) {
     if (perf_trace_enabled) {
       selector_statistics_collector.EndCollectionForCurrentRule();
       selector_statistics_collector.BeginCollectionForRule(&rule_data);
     }
-    if (can_use_fast_reject_ &&
-        selector_filter_.FastRejectSelector(
-            rule_data.DescendantSelectorIdentifierHashes(
-                rule_set->BloomHashBacking()))) {
+    if (rule_data.RejectElement(element_filter) ||
+        (can_use_fast_reject_ &&
+         selector_filter_.FastRejectSelector(
+             rule_data.DescendantSelectorIdentifierHashes(
+                 rule_set->BloomHashBacking())))) {
       fast_rejected++;
       if (perf_trace_enabled) {
         selector_statistics_collector.SetWasFastRejected();
--- a/third_party/blink/renderer/core/css/resolver/selector_filter_parent_scope_test.cc
+++ b/third_party/blink/renderer/core/css/resolver/selector_filter_parent_scope_test.cc
@@ -66,11 +66,13 @@
 
       for (const CSSSelector* selector = selectors->First(); selector;
            selector = CSSSelectorList::Next(*selector)) {
+        Element::TinyBloomFilter subject_filter = 0;
         Vector<uint16_t> selector_hashes;
         filter.CollectIdentifierHashes(*selector, /* style_scope */ nullptr,
-                                       selector_hashes);
+                                       selector_hashes, subject_filter);
         EXPECT_NE(selector_hashes.size(), 0u);
         EXPECT_FALSE(filter.FastRejectSelector(selector_hashes));
+        EXPECT_EQ(subject_filter, 0u);
       }
     }
   }
@@ -100,11 +102,13 @@
 
   for (const CSSSelector* selector = selectors->First(); selector;
        selector = CSSSelectorList::Next(*selector)) {
+    Element::TinyBloomFilter subject_filter = 0;
     Vector<uint16_t> selector_hashes;
     filter.CollectIdentifierHashes(*selector, /* style_scope */ nullptr,
-                                   selector_hashes);
+                                   selector_hashes, subject_filter);
     EXPECT_NE(selector_hashes.size(), 0u);
     EXPECT_FALSE(filter.FastRejectSelector(selector_hashes));
+    EXPECT_EQ(subject_filter, 0u);
   }
 }
 
@@ -175,10 +179,12 @@
   for (const CSSSelector* selector = selectors->First(); selector;
        selector = CSSSelectorList::Next(*selector)) {
     Vector<uint16_t> selector_hashes;
+    Element::TinyBloomFilter subject_filter = 0;
     filter.CollectIdentifierHashes(*selector, /* style_scope */ nullptr,
-                                   selector_hashes);
+                                   selector_hashes, subject_filter);
     EXPECT_NE(selector_hashes.size(), 0u);
     EXPECT_FALSE(filter.FastRejectSelector(selector_hashes));
+    EXPECT_EQ(subject_filter, 0u);
   }
 }
 
--- a/third_party/blink/renderer/core/css/rule_set.cc
+++ b/third_party/blink/renderer/core/css/rule_set.cc
@@ -187,7 +187,7 @@
   }
   bloom_hash_pos_ = bloom_hash_backing.size();
   SelectorFilter::CollectIdentifierHashes(Selector(), style_scope,
-                                          bloom_hash_backing);
+                                          bloom_hash_backing, subject_filter_);
 
   // The clamp here is purely for safety; a real rule would never have
   // as many as 255 descendant selectors.
--- a/third_party/blink/renderer/core/css/rule_set.h
+++ b/third_party/blink/renderer/core/css/rule_set.h
@@ -42,6 +42,7 @@
 #include "third_party/blink/renderer/core/css/style_rule_font_feature_values.h"
 #include "third_party/blink/renderer/core/css/style_rule_font_palette_values.h"
 #include "third_party/blink/renderer/core/css/style_rule_view_transition.h"
+#include "third_party/blink/renderer/core/dom/element.h"
 #include "third_party/blink/renderer/platform/heap/collection_support/heap_hash_map.h"
 #include "third_party/blink/renderer/platform/heap/collection_support/heap_hash_set.h"
 #include "third_party/blink/renderer/platform/heap/collection_support/heap_linked_stack.h"
@@ -162,6 +163,10 @@
                                Vector<uint16_t>& new_backing,
                                unsigned new_position);
 
+  bool RejectElement(Element::TinyBloomFilter element_filter) const {
+    return (element_filter & subject_filter_) != subject_filter_;
+  }
+
   void Trace(Visitor*) const;
 
   // This number is picked fairly arbitrary. If lowered, be aware that there
@@ -193,6 +198,14 @@
   // remember to adjust the clamping in ComputeBloomFilterHashes() too.
   unsigned bloom_hash_size_ : 8;
   unsigned bloom_hash_pos_ : 24;
+
+  // A Bloom filter that this selector needs the element to match;
+  // in other words, similar to bloom_hash_{pos_size} but for the element
+  // itself instead of the ancestor chain. This is only really useful
+  // if the selector's subject consists of multiple selectors,
+  // e.g. something like .foo.bar[baz]; otherwise, it will be redundant
+  // with bucketing.
+  Element::TinyBloomFilter subject_filter_ = 0;
 };
 
 }  // namespace blink
@@ -207,6 +220,7 @@
   unsigned b;
   unsigned c;
   unsigned d;
+  unsigned e;
 };
 
 ASSERT_SIZE(RuleData, SameSizeAsRuleData);
--- a/third_party/blink/renderer/core/css/selector_checker-inl.h
+++ b/third_party/blink/renderer/core/css/selector_checker-inl.h
@@ -169,7 +169,7 @@
                !element->ClassNames().Contains(selector->Value()))
             << element << " should have matched class " << selector->Value()
             << ", Bloom bits on element are "
-            << element->AttributeOrClassBloomFilterForDebug();
+            << element->AttributeOrClassBloomFilter();
 #endif
         return false;
       }
@@ -231,7 +231,7 @@
       DCHECK(element.CouldHaveAttribute(attr))
           << element << " should have contained attribute " << attr
           << ", Bloom bits on element are "
-          << element.AttributeOrClassBloomFilterForDebug();
+          << element.AttributeOrClassBloomFilter();
 #endif
       return attribute_item.Value() == value ||
              (case_insensitive &&
--- a/third_party/blink/renderer/core/css/selector_checker.cc
+++ b/third_party/blink/renderer/core/css/selector_checker.cc
@@ -899,7 +899,7 @@
     DCHECK(element.CouldHaveAttribute(selector_attr))
         << element << " should have contained attribute " << selector_attr
         << ", Bloom bits on element are "
-        << element.AttributeOrClassBloomFilterForDebug();
+        << element.AttributeOrClassBloomFilter();
 #endif
 
     if (AttributeValueMatches(attribute_item, match, selector_value,
@@ -959,7 +959,7 @@
                !element.ClassNames().Contains(selector.Value()))
             << element << " should have matched class " << selector.Value()
             << ", Bloom bits on element are "
-            << element.AttributeOrClassBloomFilterForDebug();
+            << element.AttributeOrClassBloomFilter();
 #endif
         return false;
       }
--- a/third_party/blink/renderer/core/css/selector_filter.cc
+++ b/third_party/blink/renderer/core/css/selector_filter.cc
@@ -192,6 +192,59 @@
   }
 }
 
+void CollectSubjectIdentifierHashes(const CSSSelector* selector,
+                                    Element::TinyBloomFilter& subject_filter) {
+  for (const CSSSelector* current = selector; current;
+       current = current->NextSimpleSelector()) {
+    switch (current->Match()) {
+      case CSSSelector::kClass:
+        if (!current->Value().empty()) {
+          subject_filter |= Element::FilterForString(current->Value());
+        }
+        break;
+      case CSSSelector::kAttributeExact:
+      case CSSSelector::kAttributeSet:
+      case CSSSelector::kAttributeList:
+      case CSSSelector::kAttributeContain:
+      case CSSSelector::kAttributeBegin:
+      case CSSSelector::kAttributeEnd:
+      case CSSSelector::kAttributeHyphen: {
+        auto attribute_name = current->Attribute().LocalName();
+        if (IsExcludedAttribute(attribute_name)) {
+          break;
+        }
+        subject_filter |= Element::FilterForAttribute(current->Attribute());
+        break;
+      }
+      case CSSSelector::kPseudoClass:
+        switch (current->GetPseudoType()) {
+          case CSSSelector::kPseudoIs:
+          case CSSSelector::kPseudoWhere:
+          case CSSSelector::kPseudoParent: {
+            // If we have a one-element :is(), :where() or &, treat
+            // as if the given list was written out as a normal subject.
+            const CSSSelector* selector_list = current->SelectorListOrParent();
+            if (selector_list &&
+                CSSSelectorList::Next(*selector_list) == nullptr) {
+              CollectSubjectIdentifierHashes(selector_list, subject_filter);
+            }
+            break;
+          }
+          default:
+            break;
+        }
+        break;
+      default:
+        break;
+    }
+
+    // Don't look past the subject.
+    if (current->Relation() != CSSSelector::kSubSelector) {
+      break;
+    }
+  }
+}
+
 }  // namespace
 
 void SelectorFilter::PushAllParentsOf(TreeScope& tree_scope) {
@@ -235,10 +288,13 @@
 void SelectorFilter::CollectIdentifierHashes(
     const CSSSelector& selector,
     const StyleScope* style_scope,
-    Vector<uint16_t>& bloom_hash_backing) {
+    Vector<uint16_t>& bloom_hash_backing,
+    Element::TinyBloomFilter& subject_filter) {
   CollectDescendantCompoundSelectorIdentifierHashes(
       selector.NextSimpleSelector(), selector.Relation(), style_scope,
       bloom_hash_backing);
+  subject_filter = 0;
+  CollectSubjectIdentifierHashes(&selector, subject_filter);
 }
 
 void SelectorFilter::Trace(Visitor* visitor) const {
--- a/third_party/blink/renderer/core/css/selector_filter.h
+++ b/third_party/blink/renderer/core/css/selector_filter.h
@@ -130,7 +130,8 @@
       const base::span<const uint16_t> identifier_hashes) const;
   static void CollectIdentifierHashes(const CSSSelector&,
                                       const StyleScope*,
-                                      Vector<uint16_t>& bloom_hash_backing);
+                                      Vector<uint16_t>& bloom_hash_backing,
+                                      Element::TinyBloomFilter& subject_filter);
 
   void Trace(Visitor*) const;
 
--- a/third_party/blink/renderer/core/css/selector_filter_test.cc
+++ b/third_party/blink/renderer/core/css/selector_filter_test.cc
@@ -6,6 +6,7 @@
 
 #include "testing/gtest/include/gtest/gtest.h"
 #include "third_party/blink/renderer/core/css/css_test_helpers.h"
+#include "third_party/blink/renderer/core/css/parser/css_parser.h"
 #include "third_party/blink/renderer/core/css/style_rule.h"
 #include "third_party/blink/renderer/core/css/style_scope.h"
 #include "third_party/blink/renderer/core/testing/page_test_base.h"
@@ -34,8 +35,9 @@
   const StyleScope* style_scope =
       scope_rule ? &scope_rule->GetStyleScope() : nullptr;
 
+  Element::TinyBloomFilter subject_filter;
   SelectorFilter::CollectIdentifierHashes(*inner_style_rule->FirstSelector(),
-                                          style_scope, result);
+                                          style_scope, result, subject_filter);
   return result;
 }
 
@@ -89,4 +91,64 @@
   EXPECT_NE(0u, hashes[2]);  // .a
 }
 
+Element::TinyBloomFilter SubjectFilterForSelector(const char* selector) {
+  HeapVector<CSSSelector> arena;
+  base::span<CSSSelector> selector_vector = CSSParser::ParseSelector(
+      MakeGarbageCollected<CSSParserContext>(
+          kHTMLStandardMode, SecureContextMode::kInsecureContext),
+      CSSNestingType::kNone, /*parent_rule_for_nesting=*/nullptr, nullptr,
+      selector, arena);
+  CSSSelectorList* selectors =
+      CSSSelectorList::AdoptSelectorVector(selector_vector);
+
+  Element::TinyBloomFilter subject_filter;
+
+  Vector<uint16_t> selector_hashes;
+  SelectorFilter::CollectIdentifierHashes(*selectors->First(),
+                                          /* style_scope */ nullptr,
+                                          selector_hashes, subject_filter);
+  return subject_filter;
+}
+
+TEST_F(SelectorFilterTest, UniversalSelectorHasNoSubjectBits) {
+  EXPECT_EQ(SubjectFilterForSelector("*"), 0u);
+  EXPECT_EQ(SubjectFilterForSelector("div *"), 0u);
+  EXPECT_EQ(SubjectFilterForSelector("div > *"), 0u);
+}
+
+TEST_F(SelectorFilterTest, ClassHasSubjectBits) {
+  EXPECT_NE(SubjectFilterForSelector(".a"), 0u);
+  EXPECT_EQ(SubjectFilterForSelector(".a"), SubjectFilterForSelector("div .a"));
+  EXPECT_EQ(SubjectFilterForSelector(".a"),
+            SubjectFilterForSelector("div > .a"));
+  EXPECT_EQ(SubjectFilterForSelector(".a"),
+            SubjectFilterForSelector("div + .a"));
+}
+
+TEST_F(SelectorFilterTest, AttributeHasSubjectBits) {
+  EXPECT_NE(SubjectFilterForSelector("[a]"), 0u);
+  // Different due to uppercasing of attributes. (The odds of being equal
+  // by accident are very low.)
+  EXPECT_NE(SubjectFilterForSelector("[a]"), SubjectFilterForSelector(".a"));
+}
+
+TEST_F(SelectorFilterTest, MultipleClassesShouldBeStrictSupersetOfSingle) {
+  EXPECT_NE(SubjectFilterForSelector(".a.b"), SubjectFilterForSelector(".a"));
+  EXPECT_EQ(SubjectFilterForSelector(".a.b") & SubjectFilterForSelector(".a"),
+            SubjectFilterForSelector(".a"));
+}
+
+TEST_F(SelectorFilterTest, IsAndWhere) {
+  EXPECT_EQ(SubjectFilterForSelector(":is(.a)"),
+            SubjectFilterForSelector(".a"));
+  EXPECT_EQ(SubjectFilterForSelector(":is(.a.b)"),
+            SubjectFilterForSelector(".a.b"));
+  EXPECT_EQ(SubjectFilterForSelector(".a:is(.b)"),
+            SubjectFilterForSelector(".a.b"));
+  EXPECT_EQ(SubjectFilterForSelector(":where(.a)"),
+            SubjectFilterForSelector(".a"));
+  EXPECT_EQ(SubjectFilterForSelector(":has(.a)"), 0u);
+  EXPECT_EQ(SubjectFilterForSelector(":is(.a,.b)"), 0u);
+}
+
 }  // namespace blink
--- a/third_party/blink/renderer/core/css/selector_query.cc
+++ b/third_party/blink/renderer/core/css/selector_query.cc
@@ -170,7 +170,7 @@
       DCHECK(!element.HasClassName(class_name))
           << element << " should have contained class " << class_name
           << ", Bloom bits on element are "
-          << element.AttributeOrClassBloomFilterForDebug();
+          << element.AttributeOrClassBloomFilter();
 #endif
       continue;
     }
@@ -320,7 +320,7 @@
       DCHECK(element.CouldHaveAttributeWithPrecomputedFilter(filter))
           << element << " should have contained attribute " << selector_attr
           << ", Bloom bits on element are "
-          << element.AttributeOrClassBloomFilterForDebug();
+          << element.AttributeOrClassBloomFilter();
 #endif
 
       if (AttributeValueMatchesExact(attribute_item, selector_value,
--- a/third_party/blink/renderer/core/dom/element.h
+++ b/third_party/blink/renderer/core/dom/element.h
@@ -778,11 +778,11 @@
   bool CouldHaveClassWithPrecomputedFilter(TinyBloomFilter filter) const {
     return (attribute_or_class_bloom_ & filter) == filter;
   }
-#if DCHECK_IS_ON()
-  TinyBloomFilter AttributeOrClassBloomFilterForDebug() const {
+  // Useful if you are to match the same element against a lot of different
+  // selectors in quick succession.
+  TinyBloomFilter AttributeOrClassBloomFilter() const {
     return attribute_or_class_bloom_;
   }
-#endif
 
   // Step 5 of https://dom.spec.whatwg.org/#concept-node-clone
   virtual void CloneNonAttributePropertiesFrom(const Element&,
